metrics2/statsd/statsd.go

```diff
@@ -0,0 +1,359 @@
+type ProviderOption func(*Provider)
+
+// WithFloatValues instructs the provider to emit values to the StatsD backend
+// as floats. Only certain StatsD servers support this mode, so check to make
+// sure. If this option isn't provided, all values will be truncated to ints.
+func WithFloatValues() ProviderOption {
+	return func(p *Provider) { p.floatValues = true }
+}
+
+// WithSampleRate instructs the provider to only record, and emit, a percentage
+// of actual observations. The primary purpose is to restrict the amount of
+// bandwidth used to transmit a report to a StatsD backend.
+func WithSampleRate(rate float64) ProviderOption {
+	if rate < 0.0 {
+		rate = 0.0
+	}
+	if rate > 1.0 {
+		rate = 1.0
+	}
+	return func(p *Provider) { p.sampleRate = rate }
+}
+
+// NewProvider returns a new, empty, idle provider. Callers must be sure to
+// invoke WriteLoop or SendLoop to actually emit information to a StatsD
+// backend. The logger is used to report transport errors.
+func NewProvider(logger log.Logger, options ...ProviderOption) *Provider {
+	p := &Provider{
+		counters:    map[string]float64{},
+		gauges:      map[string]float64{},
+		timers:      map[string][]float64{},
+		histograms:  map[string][]float64{},
+		floatValues: false,
+		sampleRate:  1.0,
+		logger:      logger,
+	}
+	for _, option := range options {
+		option(p)
+	}
+	return p
+}
+
+// SendLoop connects to a StatsD backend on the given network and address, and
+// emits a report every time the passed channel fires. For typical usage, create
+// a time.NewTicker and pass the ticker.C channel to this function. The channel
+// blocks until the passed channel is closed.
+func (p *Provider) SendLoop(c <-chan time.Time, network, address string) {
+	p.WriteLoop(c, conn.NewDefaultManager(network, address, p.logger))
+}
+
+// WriteLoop writes a report to the passed writer every time the passed channel
+// fires. For typical usage, create a time.NewTicker and pass the ticker.C
+// channel to this function. The channel blocks until the passed channel is
+// closed.
+//
+// This is a low-level function, primarily intended for testing. Most callers
+// should prefer SendLoop.
+func (p *Provider) WriteLoop(c <-chan time.Time, w io.Writer) {
+	for range c {
+		if _, err := p.WriteTo(w); err != nil {
+			p.logger.Log("err", err)
+		}
+	}
+}
+
+// WriteTo flushes the buffered contents of the metrics to the passed writer, in
+// StatsD format. WriteTo is best-effort and fails fast; observations are lost
+// if there's a problem with the write. Clients should be sure to call WriteLoop
+// regularly, ideally through the SendLoop or WriteLoop helper methods.
+//
+// This is a low-level function, primarily intended for testing. Most callers
+// should prefer SendLoop.
+func (p *Provider) WriteTo(w io.Writer) (int64, error) {
+	// Copy the maps and reset them to empty.
+	// Do this in a closure to minimize lock time.
+	c, g, t, h := func() (c, g map[string]float64, t, h map[string][]float64) {
```

Comment:
The point is that you only hold the lock to copy the data out, not when you're writing the data to the remote.